Mathlib/GroupTheory/SpecificGroups/Coxeter.lean

```diff
@@ -0,0 +1,357 @@
+/-
+Copyright (c) 2023 Newell Jensen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Newell Jensen
+-/
+import Mathlib.Data.ENat.Basic
+import Mathlib.Data.Matrix.Notation
+import Mathlib.GroupTheory.PresentedGroup
+import Mathlib.LinearAlgebra.Matrix.Symmetric
+
+/-!
+# Coxeter Groups
+
+This file defines Coxeter Groups.
+
+It introduces a type class `IsCoxeterGroup` for groups that are isomorphic to a group
+presentation corresponding to a Coxeter matrix.
+
+The finite Coxeter groups are classified as the families:
+
+* Aₙ, Bₙ, Dₙ, I₂ₘ
+
+And the six exceptional groups:
+
+* E₆, E₇, E₈, F₄, H₃, H₄
+
+## Main definitions
+
+* `Matrix.IsCoxeter` : A matrix `IsCoxeter` if it is a `B × B` symmetric matrix with ones on
+  the diagonal and off-diagonal elements are greater than or equal to 2.
+* `Matrix.ToCoxeterPresentation` : The group presentation corresponding to a Coxeter matrix.
+* `CoxeterGroupBasis` : A structure recording the isomorphism between a group `W` and the group
+  presentation corresponding to a Coxeter matrix, i.e. `Matrix.ToCoxeterPresentation B M`.
+* `IsCoxeterGroup` : A group is a Coxeter group if it admits a Coxeter group basis.
+
+## References
+
+* https://en.wikipedia.org/wiki/Coxeter_group
+
+* [N. Bourbaki, *Lie Groups and Lie Algebras, Chapters 4--6*](bourbaki1968) chapter IV
+  pages 13--15
+
+## Tags
+
+coxeter group basis, coxeter group
+-/
+
+noncomputable section
+
+variable {B : Type*} [DecidableEq B] [Fintype B] [HPow (FreeGroup B) ℕ∞ (FreeGroup B)]
+
+variable (M : Matrix B B ℕ∞)
+
+variable (B)
+
+/-- A Coxeter matrix is a B × B symmetric matrix with ones on the
+diagonal and off-diagonal elements greater than or equal to 2. -/
+def Matrix.IsCoxeter (M : Matrix B B ℕ∞) : Prop :=
+    (∀ i j : B, (i = j → M i j = 1) ∧ (i ≠ j → 2 ≤ M i j)) ∧ M.IsSymm
+
+namespace CoxeterGroup
+
+open ENat
+
+variable {B}
+
+namespace Relations
+
+/-- The relation terms corresponding to a Coxeter matrix. -/
+def ofMatrix : B × B → FreeGroup B :=
+  Function.uncurry fun i j =>
+    (FreeGroup.of i * FreeGroup.of j) ^ M i j
+
+/-- The relations corresponding to a Coxeter matrix. -/
+def toSet : Set (FreeGroup B) :=
+  Set.range <| ofMatrix M
+
+end Relations
+
+end CoxeterGroup
+
+/-- The group presentation corresponding to a Coxeter matrix.
+
+Note that it is defined for any matrix of natural numbers. Its value for non-Coxeter
+matrices should be regarded as junk. `IsCoxeterGroup` checks that the matrix `IsCoxeter`. -/
+abbrev Matrix.ToCoxeterPresentation := PresentedGroup <| CoxeterGroup.Relations.toSet M
```

Comment:
`CoxeterGroup` is a better name than `ToCoxeterPresentation`: `PresentedGroup` is a type, the type of elements in the group given by the presentation...